### PR TITLE
Added data_format to flatten

### PR DESCRIPTION
**Pull request by joeyearsley**
If you train a model in NCHW format but then do inference in NHWC the results are different as the reshape sets up the resulting array differently.

### CODE REVIEW

**Comment by fchollet** [comment]
That's a reasonable change, and we can merge it. Please replicate it in `tensorflow/python/keras/_impl/keras/layers` and in the repo `keras-team/keras/` as well. Please also add a unit test for the behavior you are trying to achieve.

**Comment by joeyearsley** [comment]
Will add to the `keras-team/keras/` repo.

I've added the arguments to `tensorflow/python/keras/_impl/keras/layers`, don't think anything else is needed as it defaults to the `tf.layers.Flatten` anyway.

**Comment by joeyearsley** [comment]
I have realised this implementation constrains the input to 4 dims.
Will change to enable `N-dims, s.t. N > 2`.

File: tensorflow/python/layers/core.py [hunk 1]
```diff
@@ -368,11 +375,19 @@ class Flatten(base.Layer):
...
     super(Flatten, self).__init__(**kwargs)
     self.input_spec = base.InputSpec(min_ndim=2)
+    self.data_format = utils.normalize_data_format(data_format)
 
   def call(self, inputs):
+    if self.data_format == 'channels_first':
+      # Ensure works for any dim, plus 2 due to [N, C, ...]
+      permutation = [0]
+      permutation.extend([i + 2 for i in
+                          range(len(inputs.shape.as_list()[2:]))])
```
Comment (fchollet) [inline tensorflow/python/layers/core.py]: `inputs.shape.as_list()` looks incorrect. It should be `inputs.get_shape().as_list()`. Also take into account the case where `inputs.get_shape()` returns None (in this case, do nothing).

Comment (joeyearsley) [reply tensorflow/python/layers/core.py]: Could catch, I should be using the getter.

I've added the check to ensure the shape is fully defined after the batch dim.
Would it be worth adding a `tf.logging.warning` mentioning the possible side-effects of running it in this way?

**Comment by joeyearsley** [comment]
@fchollet Any more to do on the TF side?

**Comment by martinwicke** [comment]
@fchollet and other comments?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @martinwicke: It has been 36 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Review by martinwicke — DISMISSED** [review]

**Comment by martinwicke** [comment]
@joeyearsley can you resolve the conflicts?

**Comment by joeyearsley** [comment]
@martinwicke No matter what I try I can't resolve the conflict with that file, is it due to that file not existing any more in the master repo? 

Would deleting that file fix this issue?

**Comment by drpngx** [comment]
@joeyearsley could you pull rebase and push again?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by joeyearsley** [comment]
Pull master and rebase my latest commit to master?

Should I roll back a commit (undo the delete) before rebasing?

**Comment by fchollet** [comment]
I think this may have already been fixed at master. What happens when you rebase?

**Comment by martinwicke** [comment]
If you can do a rebase, that may be easier, since we already have all the context here.

**Comment by joeyearsley** [comment]
Hopefully that should be done correctly @martinwicke

**Review by martinwicke — DISMISSED** [review]

**Comment by drpngx** [comment]
```
FAIL: Found 2 non-whitelited pylint errors:
tensorflow/python/layers/core_test.py:477: [C0301(line-too-long), ] Line too long (85/80)
tensorflow/python/layers/core_test.py:493: [C0301(line-too-long), ] Line too long (81/80)
```

**Comment by joeyearsley** [comment]
Resolved @drpngx

**Comment by martinwicke** [comment]
Sanity build failure is odd. I'll rerun.

**Comment by martinwicke** [comment]
Ok, can you update the golden files? That'll fix the API test (instructions in the test log). 

Also, the layers test is failing, can you take a look?

**Comment by joeyearsley** [comment]
I've fixed the tests now. 

I've ran the command attached to the log however I cannot tell if it has updated them or not.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @martinwicke: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by karmel** [comment]
It looks like the API tests are still failing-- did you run the scripts detailed at https://github.com/tensorflow/tensorflow/tree/master/tensorflow/tools/api/tests ? You will need to commit the changed files to this PR.

**Comment by drpngx** [comment]
Double-checking with @martinwicke for API review.

**Comment by martinwicke** [comment]
Good for API review

**Comment by joeyearsley** [comment]
@karmel I followed the instructions provided, however I can't tell if it worked or not.

Meanwhile for some reason the other tests failing are:
- an XLA op test which I've not touched
- an timeout in `tf.lite`

**Comment by karmel** [comment]
The API tests should modify a some files (`tensorflow/tools/api/golden/tensorflow.pbtxt` or similar), which then must be git committed to the fork you are working on here. Do you see those files being modified?

**Comment by joeyearsley** [comment]
This is the only output I get:

> bazel-bin/tensorflow/tools/api/tests/api_compatibility_test --update_goldens true
sWARNING:tensorflow:VARIABLES collection name is deprecated, please use GLOBAL_VARIABLES instead; VARIABLES will be removed after 2017-03-02.
> ..
> ----------------------------------------------------------------------
> Ran 3 tests in 0.383s
>
>OK (skipped=1)

And no updated protobufs are output.

Any further ideas? - I don't understand why it's failing to update them

**Comment by drpngx** [comment]
You have to run the program with python2, otherwise the test will be skipped. This is so that we have a stable interface.

**Comment by drpngx** [comment]
Oh, you need to pull rebase and push again to resolve the conflict.

**Comment by drpngx** [comment]
Could you check https://source.cloud.google.com/results/invocations/452569d4-59c2-407c-875a-29752a47122e/log ? (look for `api_compatibility_test`)

In the diff `broadcast_to` seems to have disappeared. Maybe you ran in between updates. Thanks.

**Comment by joeyearsley** [comment]
@drpngx Seems to be the case!

**Comment by martinwicke** [comment]
@fchollet what is the state of this?

**Comment by joeyearsley** [comment]
@fchollet Any update? - This was merged into Keras months ago, I would really like to get it off my plate.

**Comment by martinwicke** [comment]
Ok, this is weird, this should have been merged. I'll look into why it's stuck.

**Comment by martinwicke** [comment]
I retriggered the tests. I think there's still a problem with the API goldens, could you take a look? It's probably as easy as applying this diff:

```
   member_method {
      name: "flatten"
-     argspec: "args=[\'inputs\', \'name\'], varargs=None, keywords=None, defaults=[\'None\'], "
+     argspec: "args=[\'inputs\', \'data_format\', \'name\'], varargs=None, keywords=None, defaults=[\'channels_last\', \'None\'], "
    }
```
to api/golden/v2/tensorflow.layers....

**Comment by joeyearsley** [comment]
This seems to be the V2 branch? 
Will this not be updated frequently anyway internally as you are still making TF v2?

**Comment by martinwicke** [comment]
It's updated whenever me make a change. 

You are correct in that this particular class is likely to not survive, but we will make this change in a separate commit (in progress). Until then, we need to keep the v2 goldens up to date, we don't regenerate them in the background.

**Comment by asimshankar** [comment]
@joeyearsley : It isn't in V2 only, it is for V1 as well. Basically, the `data_format` argument should be the last one (after `name`), otherwise existing callers to `flatten()` that use positional arguments will break (e.g., `tf.layers.flatten(x, 'foo')`)

**Comment by joeyearsley** [comment]
@martinwicke please can you re-run the tests before something else changes

**Comment by asimshankar** [comment]
@joeyearsley : Did you see https://github.com/tensorflow/tensorflow/pull/17672#issuecomment-426088246 ? The tests will still fail since the ordering of the `name` argument has changed. Could you update that?

**Comment by joeyearsley** [comment]
Ok, now I've updated to match the ordering - in future please comment on a code review so I can see exactly what line etc.

However, if it's a kwarg it should probably be named in the test also.

Now please can we get this sorted, 6 months is a bit long when it should have been updated when the code was refactored to rely on Keras' flatten layer - which has the `data_format` kwarg since I added that at the start of this PR.
